nixos/modules/services/backup/bupstash.nix

```diff
@@ -0,0 +1,666 @@
+{
+  config,
+  lib,
+  pkgs,
+  utils,
+  ...
+}:
+
+let
+  inherit (lib)
+    concatLists
+    concatMap
+    concatMapAttrs
+    getExe
+    hasPrefix
+    literalExpression
+    mapAttrs'
+    mapAttrsToList
+    mkDefault
+    mkEnableOption
+    mkIf
+    mkMerge
+    mkOption
+    mkPackageOption
+    nameValuePair
+    optional
+    optionalAttrs
+    optionalString
+    ;
+
+  inherit (lib.types)
+    attrsOf
+    bool
+    either
+    enum
+    lines
+    listOf
+    nullOr
+    passwdEntry
+    path
+    str
+    submodule
+    ;
+
+  inherit (utils) escapeSystemdExecArgs;
+
+  mkJobTarget = name: _: {
+    name = "bupstash-${name}";
+    value = {
+      description = "Bupstash group job for ${name}.";
+      unitConfig = {
+        StopWhenUnneeded = true;
+      };
+    };
+  };
+
+  mkJobTimer = name: job: {
+    name = "bupstash-${name}";
+    value = {
+      wantedBy = [ "timers.target" ];
+      timerConfig = {
+        Unit = "bupstash-${name}.target";
+        OnCalendar = job.startAt;
+      };
+    };
+  };
+
+  mkJobUnits =
+    name: job:
+    {
+      "bupstash-${name}-prehook" = mkIf (job.preHook != "") {
+        description = "Hook to run before the ${name} bupstash backup.";
+
+        partOf = [ "bupstash-${name}.target" ];
+        wantedBy = [ "bupstash-${name}.target" ];
+
+        script = job.preHook;
+
+        serviceConfig = {
+          # Prevents the unit to run multiple times before the postHook has finished
+          RemainAfterExit = true;
+          StateDirectory = "bupstash";
+          Type = "oneshot";
+          WorkingDirectory = "/var/lib/bupstash";
+        };
+      };
+
+      "bupstash-${name}-posthook" = mkMerge [
+        (mkIf (job.postHook != "") {
```

Comment:
You cannot set a script to the empy string for systemd services, as it will not create an `ExecStart` entry, and thens systemd will log an error, so we need to filter out empty hooks anyway

Reply:
The reference for this is https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/lib/systemd-unit-options.nix#L392 so nixpkgs already considers that the empty strings means no script.